tests/integration/test_integration.py

```diff
@@ -29,209 +31,351 @@
+k8s = KubernetesClient()
 
 
 class TestVaultK8s:
-    @staticmethod
-    async def deploy_charm(ops_test: OpsTest, charm: Path) -> None:
-        """Deploys charm.
-
-        Args:
-            ops_test: Ops test Framework.
-            charm: Charm path.
-        """
+    @pytest.fixture(scope="module")
+    async def deploy_phase_1(self, ops_test: OpsTest):
+        copy_lib_content()
+        resources = {"vault-image": METADATA["resources"]["vault-image"]["upstream-source"]}
         assert ops_test.model
-        resources = {
-            "vault-image": METADATA["resources"]["vault-image"]["upstream-source"],
-        }
-        await ops_test.model.deploy(
-            charm,
+
+        built_charms = await ops_test.build_charms(".", f"{VAULT_KV_REQUIRER_CHARM_DIR}/")
+        vault_charm = built_charms.get("vault-k8s", "")
+        vault_kv_requirer_charm = built_charms.get("vault-kv-requirer", "")
+
+        deploy_vault = ops_test.model.deploy(
+            vault_charm,
             resources=resources,
             application_name=APPLICATION_NAME,
             trust=True,
             series="jammy",
             num_units=NUM_VAULT_UNITS,
             config={"common_name": "example.com"},
         )
+        deploy_traefik = ops_test.model.deploy(
+            "traefik-k8s",
+            application_name=TRAEFIK_APPLICATION_NAME,
+            trust=True,
+            channel="stable",
+        )
+        deploy_s3_integrator = ops_test.model.deploy(
+            "s3-integrator",
+            application_name=S3_INTEGRATOR_APPLICATION_NAME,
+            trust=True,
+        )
+        deploy_minio = ops_test.model.deploy(
+            "minio", application_name=MINIO_APPLICATION_NAME, trust=True, config=MINIO_CONFIG
+        )
+        deploy_vault_kv_requirer = ops_test.model.deploy(
+            vault_kv_requirer_charm,
+            application_name=VAULT_KV_REQUIRER_APPLICATION_NAME,
+            num_units=1,
+        )
+        await asyncio.gather(
+            deploy_s3_integrator,
+            deploy_minio,
+            deploy_vault_kv_requirer,
+            deploy_vault,
+            deploy_traefik,
+        )
+        await ops_test.model.wait_for_idle(
+            apps=[S3_INTEGRATOR_APPLICATION_NAME],
+            status="blocked",
+            timeout=600,
+            wait_for_exact_units=1,
+        )
+        await ops_test.model.wait_for_idle(
+            apps=[
+                MINIO_APPLICATION_NAME,
+                VAULT_KV_REQUIRER_APPLICATION_NAME,
+                TRAEFIK_APPLICATION_NAME,
+            ],
+            status="active",
+            timeout=600,
+            wait_for_exact_units=1,
+        )
+        await ops_test.model.wait_for_idle(
+            apps=[APPLICATION_NAME],
+            status="active",
+            timeout=600,
+            wait_for_exact_units=NUM_VAULT_UNITS,
+        )
+        yield
```

Comment:
Why is this needed?

Reply:
The yield is where all of the phase 1 tests run. Any line afterward is to clean up the fixture after all tests that depend on phase 1 is complete.

Reply:
Do we really need this? We are making the integration tests less readable through this PR. Let's make sure it's justified and we only do what really matters to address the performance issue we're observing

Reply:
The yield keyword is explicitly called out for teardowns in the [pytest documentation](https://docs.pytest.org/en/7.1.x/how-to/fixtures.html#teardown-cleanup-aka-fixture-finalization). It's being used here exactly as intended by pytest.

I disagree that this is less readable. Having fixtures contain their definition and teardown logic in the same function and be marked appropriately as fixtures is much clearer than having multiple different deploy functions scattered across the file, some marked as static methods when they should be fixtures, and are called in a specific order which makes the tests implicitly depend on each other.

Reply:
I agree with Kayra for the yield keyword. This is the proper way to do cleanup code with fixtures in Pytest.